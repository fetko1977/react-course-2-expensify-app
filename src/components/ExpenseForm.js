import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import Select from 'react-select';
import categories from '../categories/categories';

class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        
        this.state = {
            category: props.expense && props.expense.categoryClass !== '' && props.expense.categoryLabel !== '' ? {value: props.expense.categoryClass, label: props.expense.categoryLabel} : null,
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onCategoryChange = (category) => {
        this.setState(() => ({ category }));
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({
                createdAt
            }))
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({
            calendarFocused: focused
        }))
    }

    submit = (e) => {
        e.preventDefault();
        const category = this.state.category === null ? {value: '', label: ''} : this.state.category;

        if (!this.state.description || !this.state.amount) {
            // Set error state  - Please provide description and amount
            this.setState(() => ({
                error: 'Please provide description and amount'
            }))
        } else {
            this.setState(() => ({
                error: ''
            }))
            this.props.onSubmit({
                categoryClass: category.value,
                categoryLabel: category.label,
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }

    render() {
        return (
            <form className="form" onSubmit={this.submit}>
                { this.state.error && <p className="form__error">{this.state.error}</p>}
                <Select
                    value={this.state.category}
                    onChange={this.onCategoryChange}
                    isClearable={true}
                    options={categories}
                    placeholder="Select Category"
                />
                <input 
                    type="text" 
                    className="text-input" 
                    placeholder="Description" 
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input 
                    type="text"
                    className="text-input" 
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea 
                    placeholder="Add note to your expense (optional)"
                    className="textarea"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                ></textarea>
                <div>
                    <button className="btn btn--blue">Save Expense</button>
                </div>
            </form>
        )
    }
}

export default ExpenseForm;