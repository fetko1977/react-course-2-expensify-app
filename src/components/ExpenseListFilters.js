import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setCategoryFilter, setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import Select from 'react-select';
import categories from '../categories/categories';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
    category: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
      this.setState(() => ({ calendarFocused }));
  }

  onCategoryChange = (category) => {
      const categoryLabel = category === null ? '' : category.label;
      this.setState(() => ({ category }));
      this.props.setCategoryFilter(categoryLabel);
  }

  onTextChange = (e) => {
      this.props.setTextFilter(e.target.value);
  }

  onSortChange = (e) => {
    if (e.target.value === "amount") {
      this.props.sortByAmount();
    } else if (e.target.value === "date") {
      this.props.sortByDate();
    }
  }
  render() {
    return (
      <div className="content-container">
        <div className="input-group">
            <div className="input-group__item">
                <input type="text" className="text-input" placeholder="Search expenses" value={this.props.filters.text} onChange={this.onTextChange} />
            </div>
            <div className="input-group__item">
                <select className="select" value={this.props.filters.sortBy} onChange={this.onSortChange}>
                  <option value="date">Date</option>
                  <option value="amount">Amount</option>
                </select>
            </div>
            <div className="input-group__item input-group__item--category-filter">
                <Select
                    value={this.state.category}
                    isClearable={true}
                    onChange={this.onCategoryChange}
                    options={categories}
                    placeholder="Filter By Category"
                />
            </div>
            <div className="input-group__item">
                <DateRangePicker
                  startDate={this.props.filters.startDate}
                  endDate={this.props.filters.endDate}
                  onDatesChange={this.onDatesChange}
                  focusedInput={this.state.calendarFocused}
                  onFocusChange={this.onFocusChange}
                  showClearDates={true}
                  numberOfMonths={1}
                  isOutsideRange={() => false}
                />
            </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({ filters: state.filters });

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    setCategoryFilter: (categoryLabel) => dispatch(setCategoryFilter(categoryLabel)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);