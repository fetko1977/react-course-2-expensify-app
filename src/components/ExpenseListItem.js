import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faListAlt } from '@fortawesome/free-solid-svg-icons'

const ExpenseListItem = ({
        id,
        categoryClass,
        categoryLabel,
        description,
        amount,
        createdAt
    }) => {
        const listItemCategoryClass = !!categoryClass ? "list-item__category " + categoryClass : "list-item__category";
        return (
            <Link className="list-item" to={`/edit/${id}`}>
                <div>
                    <h3 className="list-item__title">{description}</h3>
                    <span className="list-item__sub-title">{moment(createdAt).format('MMMM Do, YYYY')}</span>
                    { categoryLabel && <h5 className={listItemCategoryClass}><strong>Category: </strong>{categoryLabel}</h5>}
                </div>
                <h3  className="list-item__data">{numeral(amount / 100).format('$0,0.00')}</h3>
            </Link>
        )
    };

export default ExpenseListItem;