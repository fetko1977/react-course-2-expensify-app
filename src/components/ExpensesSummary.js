import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => {
    const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
    return (
        <div>
            <h5>Viewing {expensesCount} {expenseWord} totalling {numeral(expensesTotal / 100).format('$0,0.00')}</h5>
        </div>
    )
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary);
