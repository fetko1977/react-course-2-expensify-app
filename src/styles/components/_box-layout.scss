.box-layout {
    align-items: center;
    background: url('/images/bg.jpg');
    background-size: cover;
    display: flex;
    justify-content: center;
    width: 100vw;
    height: 100vh;
}

.box-layout__box {
    background: fade-out(white, .15);
    border-radius: 3px;
    text-align: center;
    width: 25rem;
    padding: $l-size $m-size;
}

.box-layout__title {
    margin: 0 0 $m-size;
    line-height: 1;
}