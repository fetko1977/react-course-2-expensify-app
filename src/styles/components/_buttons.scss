.btn {
    font-size: 16px;
    border: none;
    line-height: 1;
    padding: $s-size;
    display: inline-block;
    text-decoration: none;
    font-size: $font-size-large;
    font-weight: 300;
    color: white;
}

.btn--blue {
    background: $blue;
    
}

.btn--link {
    background: transparent;
}

.btn--secondary {
    background: #888;
}