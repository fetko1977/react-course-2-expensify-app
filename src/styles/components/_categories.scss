
.list-item__category {
    &.car {
        background: lighten(orange, 10%);
    }

    &.education {
        background: lighten(#fa2afa, 10%);
    }

    &.groceries {
        background: lighten(rgb(55, 255, 55), 10%);
    }

    &.rent {
        background: lighten(rgb(2, 19, 255), 10%);
    }

    &.other {
        background: lighten(rgb(46, 46, 46), 10%);
    }
}