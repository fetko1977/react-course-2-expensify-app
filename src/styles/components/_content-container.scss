.content-container {
    max-width: 80vw;
    padding: 0 $m-size;
    margin: 0 auto;
}