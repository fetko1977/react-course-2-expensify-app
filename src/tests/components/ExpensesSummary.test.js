import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should render ExpensesSummary with 1 expense', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={232}/>);

    expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={32} expensesTotal={435500}/>);

    expect(wrapper).toMatchSnapshot();
});

