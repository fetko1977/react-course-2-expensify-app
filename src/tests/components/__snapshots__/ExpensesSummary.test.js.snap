// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with 1 expense 1`] = `
<div>
  <h5>
    Viewing 
    1
     
    expense
     totalling 
    $2.32
  </h5>
</div>
`;

exports[`should render ExpensesSummary with multiple expenses 1`] = `
<div>
  <h5>
    Viewing 
    32
     
    expenses
     totalling 
    $4,355.00
  </h5>
</div>
`;
